service/matching/matcher_test.go

```diff
@@ -210,55 +210,67 @@ func (t *MatcherTestSuite) testRemoteSyncMatch(taskSource enumsspb.TaskSource) {
+		_, err := t.rootMatcher.Poll(ctx, &pollMetadata{})
+		t.Assert().NoError(err)
+		cancel()
+	}()
+	time.Sleep(2 * time.Millisecond) //nolint:forbidigo
+
+	go func() {
+		historyTask.responseC <- nil
+	}()
+	time.Sleep(2 * time.Millisecond) //nolint:forbidigo
+	happened, err := t.rootMatcher.Offer(ctx, historyTask)
+	t.True(happened)
+	t.Nil(err)
+}
+
 func (t *MatcherTestSuite) TestRejectSyncMatchWhenBacklog() {
 	historyTask := newInternalTaskForSyncMatch(randomTaskInfo().Data, nil)
-	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	ctx, cancel := context.WithTimeout(context.Background(), 1*time.Second)
 
 	// sync match happens when there is no backlog
-	t.client.EXPECT().PollWorkflowTaskQueue(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.PollWorkflowTaskQueueResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.Poll(ctx, &pollMetadata{}) //nolint:errcheck
-	time.Sleep(time.Millisecond)
-	go func() { historyTask.responseC <- nil }()
-	happened, err := t.childMatcher.Offer(ctx, historyTask)
-	t.True(happened)
-	t.Nil(err)
+	t.validateSyncMatchWhenNoBacklog(ctx, historyTask)
 
+	// first task waits for a local poller
 	intruptC := make(chan struct{})
 	youngBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Second), nil)
-	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
-	go t.childMatcher.MustOffer(ctx, youngBacklogTask, intruptC) //nolint:errcheck
+	go t.rootMatcher.MustOffer(ctx, youngBacklogTask, intruptC) //nolint:errcheck
 	time.Sleep(time.Millisecond)
 
 	// should allow sync match when there is only young tasks in the backlog
-	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
-	happened, err = t.childMatcher.Offer(ctx, historyTask)
+	t.client.EXPECT().AddWorkflowTask(gomock.Any(), gomock.Any(), gomock.Any()).Do(
+		func(arg0 context.Context, arg1 *matchingservice.AddWorkflowTaskRequest, arg2 ...interface{}) {
+		},
+	).Return(&matchingservice.AddWorkflowTaskResponse{}, errMatchingHostThrottleTest)
+	happened, err := t.childMatcher.Offer(ctx, historyTask)
 	t.Nil(err)
 	t.False(happened) // sync match did not happen, but we called the forwarder client
 
-	// should not allow sync match when there is an old task in backlog
+	// second task waits for a local poller
 	oldBacklogTask := newInternalTaskFromBacklog(randomTaskInfoWithAge(time.Minute), nil)
-	go t.childMatcher.MustOffer(ctx, oldBacklogTask, intruptC) //nolint:errcheck
+	go t.rootMatcher.MustOffer(ctx, oldBacklogTask, intruptC) //nolint:errcheck
 	time.Sleep(time.Millisecond)
-	happened, err = t.childMatcher.Offer(ctx, historyTask)
+
+	// should not allow sync match when there is an old task in backlog
+	happened, err = t.rootMatcher.Offer(ctx, historyTask)
 	t.False(happened)
```

Comment:
good point - I had initially thought about this and had decided to leave as is given that I verified, with print statements, that we were returning a `false` from the if statement testing a non-negligible backlog and not because of a lack of pollers.

Having revisited this, I felt a little uncomfortable imagining this shall get shipped. I changed this to now verify if we indeed are getting a rejection because of a non-negligible backlog and not because of no pollers waiting. I have added a explanatory comment in the test explaining my thought process.